### PR TITLE
Upgrade to FEniCS 2019.1

### PR DESCRIPTION
**Pull request by uvilla**
Upgrade `hIPPYlib` to `FEniCS 2019.1` and `pybind11`.

This version of `hIPPYlib` works with a patched version of `FEniCS 2019.1`, because some `pybind` wrappers were missing.

docker:
```
docker pull hippylib/fenics
```
anaconda:
```
conda create -n fenics-2019.1 -c uvilla -c conda-forge fenics==2019.1.0 matplotlib scipy jupyter
```

A few hints to migrate your code to `FEniCS 2019.1`:

- If you are creating a `dl.Expression` from a c++ string, make sure to add the `std::` namespace when needed.

- Use `dl.MPI.comm_world` instead of `dl.mpi_comm_world()`, and `dl.MPI.comm_self` instead of `dl.mpi_comm_self()`

- `dl.interactive` does not exists anymore use `plt.show()` (with `plt = matplotlib.pyplot`) instead.

- Use `dl.MeshFunction(...,...,mesh.geometry().dim())` instead of `dl.CellFunction(...,...)`

- Use `dl.MeshFunction(...,...,mesh.geometry().dim()-1)` instead of `dl.FacetFunction(...,...)`

- Some `ufl` functions are not wrapped by `dolfin` anymore. Import `ufl` directly when needed

- FEniCS 2019.1 always returns a `mpi4py` communicator

- If you are writing a pure python `dl.Expression` class (which is extremely slow and not reccomended), do not inherit from `dl.Expression`, instead inherit from `dl.UserExpression`

Changes in `hIPPYlib`:

- `code_AnisTensor2D` has been succeeded by `ExpressionModule.AnisTensor2D`

- All deprecated methods have been removed.

### CODE REVIEW

File: hippylib/algorithms/cpp_multivector/multivector.cpp [hunk 1]
```diff
@@ -155,6 +160,89 @@ MultiVector::~MultiVector()
...
+		.def("__len__", &hippylib::MultiVector::nvec,
+			 "The length of a multivector is the number of vector it contains")
+		.def("__getitem__", (std::shared_ptr<const dolfin::GenericVector> (hippylib::MultiVector::*)(int) const) &hippylib::MultiVector::operator[] )
+		.def("__setitem__", (std::shared_ptr<dolfin::GenericVector> (hippylib::MultiVector::*)(int)) &hippylib::MultiVector::operator[] )
+		.def("setSizeFromVector", &hippylib::MultiVector::setSizeFromVector,
+			 "Initialize a multivector by providing a vector v as template and the number of vectors nvec",
+			 py::arg("v"), py::arg("nvec"))
+		.def("dot", [](const hippylib::MultiVector & self, const dolfin::GenericVector & v)
+				{
+    				int size = self.nvec();
```
Comment (eldarkh) [inline hippylib/algorithms/cpp_multivector/multivector.cpp]: `unsigned int` if you decide to change them all

File: hippylib/algorithms/cpp_multivector/multivector.cpp [hunk 2]
```diff
@@ -155,6 +160,89 @@ MultiVector::~MultiVector()
...
+    				return ma;
+				},
+				"Perform the inner product with a vector v",
+				py::arg("v")
+    	     )
+		.def("dot", [](const hippylib::MultiVector& self, const hippylib::MultiVector & other)
+				{
+    				int size1 = self.nvec();
+    				int size2 = other.nvec();
+    				py::array_t<double> ma(size1*size2);
```
Comment (eldarkh) [inline hippylib/algorithms/cpp_multivector/multivector.cpp]: same `unsigned int` comment if to be changed

File: hippylib/algorithms/linalg.py [hunk 3]
```diff
@@ -24,74 +25,84 @@ def amg_method(amg_type="ml_amg"):
...
     
 def GetFromOwnedGid(v, gid):
-    s = cpp_module.cpp_linalg()
-    return s.GetFromOwnedGid(v, gid)
+    #    assert(v.owns_index(gid));
+    #    double val;
+    #    la_index index = static_cast<la_index>(gid);
+    #    v.get(&val, 1, &index);
+    #
+    #    return val;
```
Comment (eldarkh) [inline hippylib/algorithms/linalg.py]: Some lines are commented out here, not sure if intentional or just forgot to delete

File: hippylib/algorithms/linalg.py [hunk 4]
```diff
@@ -24,74 +25,84 @@ def amg_method(amg_type="ml_amg"):
...
 
 def SetToOwnedGid(v, gid, val):
-    s = cpp_module.cpp_linalg()
-    s.SetToOwnedGid(v, gid, val)
+    #    assert(v.owns_index(gid));
+    #    la_index index = static_cast<la_index>(gid);
+    #    v.set(&val, 1, &index);
+    # la_index_dtype
+    v[gid] = val
+
```
Comment (eldarkh) [inline hippylib/algorithms/linalg.py]: Some lines are commented out here, not sure if intentional or just forgot to delete

File: hippylib/modeling/cpp_AssemblePointwiseObservation/AssemblePointwiseObservation.cpp [hunk 5]
```diff
@@ -41,7 +47,7 @@ PetscInt PointwiseObservation::computeLGtargets(MPI_Comm comm,
 	 std::vector<int> tmp(nTargets);
 	 for(int i = 0; i < nTargets; ++i)
 	 {
-		 Point p(gdim, targets.data()+i*gdim);
+		 dolfin::Point p(gdim, targets.data()+i*gdim);
 		 if(bbt->collides_entity(p))
```
Comment (eldarkh) [inline hippylib/modeling/cpp_AssemblePointwiseObservation/AssemblePointwiseObservation.cpp]: ~~Isn't `collides_entity(p)` deprecated and removed starting from 2018.1.0?~~ Apparently it is only removed from python API

File: hippylib/modeling/expression.py [hunk 6]
```diff
@@ -106,16 +113,42 @@ class Mollifier : public Expression
...
+    
   private:
+    double l;
+    double o;
+  
+    double theta0;
+    double theta1;
+    double alpha;
+    
     int nlocations;
```
Comment (eldarkh) [inline hippylib/modeling/expression.py]: `unsigned int` comment

File: hippylib/algorithms/cpp_multivector/multivector.cpp [hunk 7]
```diff
@@ -12,19 +12,24 @@
...
+namespace hippylib
 {
 
 MultiVector::MultiVector()
 {
 }
 
-MultiVector::MultiVector(const GenericVector & v, int nvec):
+MultiVector::MultiVector(const dolfin::GenericVector & v, int nvec):
 		mv(nvec)
```
[the remark this answers is not in the document]
Comment (uvilla) [reply hippylib/algorithms/cpp_multivector/multivector.cpp]: @eldarkh ,

I'm not sure I understand your comment. Each column of the multivector should be a new vector.
Doesn't your solution just create `nvec` pointers all pointing to the same object?

File: hippylib/algorithms/linalg.py (hunk 3 above)
Comment (uvilla) [reply hippylib/algorithms/linalg.py]: You're right. I'll get rid of them.

File: hippylib/algorithms/linalg.py (hunk 4 above)
Comment (uvilla) [reply hippylib/algorithms/linalg.py]: You're right. I'll get rid of them.

File: hippylib/algorithms/cpp_multivector/multivector.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (eldarkh) [reply hippylib/algorithms/cpp_multivector/multivector.cpp]: Oops, that's my bad, you are right

File: hippylib/algorithms/cpp_multivector/multivector.hpp [hunk 8]
```diff
@@ -16,63 +16,53 @@
...
 
 class MultiVector
 {
 public:
 	MultiVector();
-	MultiVector(const GenericVector & v, int nvec);
+	MultiVector(const dolfin::GenericVector & v, int nvec);
 	MultiVector(const MultiVector & orig);
 
 	int nvec() const {return mv.size();}
```
[the remark this answers is not in the document]
Comment (uvilla) [reply hippylib/algorithms/cpp_multivector/multivector.hpp]: You are right. I think also FEniCS uses `std::size_t` to return lengths of objects.

**Pull request merged**